### ISSUE TITLE
Potential soundness issue in `oxc_syntax` 0.130.0

### ISSUE THREAD

**Issue by varies-rust** [oxc-project/oxc#22592]
Hello, and thank you for maintaining this crate.

I found a sanitizer/Miri failure reachable through public crate APIs using safe Rust code. I may be missing crate-specific preconditions, but the behavior looks worth checking because safe callers should not be able to trigger undefined behavior.

## Summary

- Crate: `oxc_syntax`
- Version tested: `0.130.0`
- API paths with similar failures observed:
  - `ReferenceId::from_usize_unchecked`
  - `ReferenceId::from_raw_unchecked`
  - `NodeId::from_usize_unchecked`
  - `NodeId::from_raw_unchecked`
  - `ScopeId::from_usize_unchecked`
  - `SymbolId::from_usize_unchecked`
  - `RedeclarationId::from_usize_unchecked`
  - `ElementId::from_usize_unchecked/from_raw_unchecked`

## Observed diagnostics

```text
Unsafe precondition violation: unsafe precondition(s) violated: NonZero::new_unchecked requires the argument to be non-zero
Miri undefined behavior: error: Undefined Behavior: entering unreachable code
```

## Reproduction

The snippets below are minimal readable reproducers. Each PoC also includes the source location most relevant to the reported failure.

### PoC 1: `test_oxc_syntax0::generated_test_0`

Relevant source location:

- `oxc_index-4.1.0/src/macros.rs:282-283`: the generated public `from_usize_unchecked` wraps `NonMaxU32::new_unchecked`.

Readable equivalent PoC:

```rust
#[test]
fn poc() {
    let _ = oxc_syntax::reference::ReferenceId::from_usize_unchecked(usize::MAX);
}
```

Observed diagnostic:

```text
unsafe precondition(s) violated: NonZero::new_unchecked requires the argument to be non-zero
```

### PoC 2: `test_oxc_syntax101::generated_test_101`

Relevant source location:

- `oxc_index-4.1.0/src/macros.rs:724-725`: primitive index types also expose unchecked construction as a safe function.
- `oxc_index-4.1.0/src/macros.rs:282-283`: converting that value into a `NonMaxU32` index reaches `new_unchecked`.

Readable equivalent PoC:

```rust
#[test]
fn poc() {
    let mut id = oxc_syntax::class::ElementId::from_usize_unchecked(0);

    id -= 1;
    let raw_index = id.index();
    let _ = oxc_syntax::node::NodeId::from_usize_unchecked(raw_index);
}
```

Observed diagnostic:

```text
error: Undefined Behavior: entering unreachable code
```

## Source review and suggested fix

Brief reasoning:
- The generated `from_usize_unchecked` constructors are public safe functions in this crate, even though they wrap `NonMaxU32::new_unchecked`.
- That makes invalid index construction a caller-visible contract bug, not just an internal helper detail.

Suggested fix:
- Make the unchecked constructors `unsafe`, or expose checked constructors that return `Option` or `Result` and keep the invariant enforced at the API boundary.

Thanks again for taking a look.

**Comment by Boshen** [comment]
These APIs are intentionally unsafe. They are used in code path that are safe for performance reasons.

**Comment by overlookmotel** [comment]
It looks like some of these APIs are not `unsafe` methods, which they should be. I believe the OP's point is valid. I will investigate further and hopefully fix.

**Comment by overlookmotel** [comment]
Fixed in https://github.com/oxc-project/oxc-index-vec/pull/179 and #23000.
